File: src/core/agent/context.ts
import type { ChatMessage } from "@/core/llm/types";
import { prisma } from "@/lib/db";
import type { AgentExecutionContext } from "./types";

const MAX_CONTEXT_MESSAGES = 50;

/**
 * Builds the conversation context for an agent working on a task.
 * Gathers relevant messages from the project/task and formats them
 * as chat messages for the LLM.
 */
export async function buildAgentContext(
  ctx: AgentExecutionContext
): Promise<ChatMessage[]> {
  const messages: ChatMessage[] = [];

  // 1. Add project context as a user message
  messages.push({
    role: "user",
    content: formatProjectContext(ctx),
  });

  // 2. Fetch relevant messages from the database
  const dbMessages = await prisma.message.findMany({
    where: {
      projectId: ctx.projectId,
      ...(ctx.taskId ? { taskId: ctx.taskId } : {}),
    },
    orderBy: { createdAt: "asc" },
    take: MAX_CONTEXT_MESSAGES,
    include: {
      sender: { include: { role: true } },
    },
  });

  // 3. Convert DB messages to chat messages
  for (const msg of dbMessages) {
    const senderLabel =
      msg.senderType === "founder"
        ? "Founder"
        : msg.senderType === "system"
          ? "System"
          : msg.sender
            ? `${msg.sender.name} (${msg.sender.role.title})`
            : "Unknown";

    if (msg.senderType === "agent" && msg.senderId === ctx.employeeId) {
      // This agent's own messages become assistant messages
      messages.push({
        role: "assistant",
        content: msg.content,
      });
    } else {
      // Other messages become user messages with sender labels
      messages.push({
        role: "user",
        content: `[${senderLabel}]: ${msg.content}`,
      });
    }
  }

  return messages;
}

/**
 * Formats the project/task context into a readable string.
 */
function formatProjectContext(ctx: AgentExecutionContext): string {
  let context = `You are ${ctx.employeeName}, serving as ${ctx.roleTitle} in this project.\n\n`;
  context += `## Project: ${ctx.projectName}\n${ctx.projectDescription}\n\n`;

  if (ctx.taskTitle && ctx.taskDescription) {
    context += `## Your Current Task: ${ctx.taskTitle}\n${ctx.taskDescription}\n\n`;
  }

  context += `## Your Workspace\n`;
  context += `You have a personal workspace (your folder in the project's Document/Code tab). Save your work there as you go:\n`;
  context += `- Use **save_to_workspace** to save drafts, outlines, research notes, and intermediate code. Call it frequently so work is persisted (e.g. path "drafts", title "outline.md").\n`;
  context += `- Use **create_file** for final deliverables to submit to the CEO. Use optional path to organize (e.g. path "src", title "Button.tsx").\n`;
  context += `- Organize with path: "docs", "drafts", "src", "research", etc. (no leading/trailing slash).\n\n`;
  context += `Please complete your assigned work. Be thorough and professional.`;

  return context;
}

/**
 * Trims messages if the total is too long, keeping the system context
 * and the most recent messages.
 */
export function trimContext(
  messages: ChatMessage[],
  maxMessages: number = MAX_CONTEXT_MESSAGES
): ChatMessage[] {
  if (messages.length <= maxMessages) return messages;

  // Always keep the first message (project context) and the last N messages
  const first = messages[0];
  const recent = messages.slice(-(maxMessages - 1));
  return [first, ...recent];
}

File: agent-engine/internal/agent/context.go
package agent

import (
	"context"
	"fmt"

	"github.com/kafkalm/bossman/agent-engine/internal/db"
	"github.com/kafkalm/bossman/agent-engine/internal/llm"
)

const maxContextMessages = 50
const projectContextMessages = 20

// BuildAgentContext assembles the message history for an agent LLM call.
// For task execution: merges recent project-level messages (context) + all task messages.
// For non-task calls: uses project-level messages only.
func BuildAgentContext(ctx context.Context, database *db.DB, emp *db.EmployeeWithRole, project *db.Project, task *db.Task) ([]llm.ChatMessage, error) {
	var messages []llm.ChatMessage

	// 1. Project context header
	messages = append(messages, llm.ChatMessage{
		Role:    "user",
		Content: formatProjectContext(emp, project, task),
	})

	if task != nil {
		// Task execution: merge project-level msgs (last 20) + all task msgs, dedup by ID
		projectMsgs, err := database.GetProjectMessages(ctx, project.ID, projectContextMessages)
		if err != nil {
			return nil, fmt.Errorf("project messages: %w", err)
		}
		taskMsgs, err := database.GetTaskMessages(ctx, task.ID)
		if err != nil {
			return nil, fmt.Errorf("task messages: %w", err)
		}

		merged := mergeAndDedup(projectMsgs, taskMsgs)
		for _, msg := range merged {
			messages = append(messages, convertDBMessage(msg, emp.ID))
		}
	} else {
		// Non-task: use recent project messages
		projectMsgs, err := database.GetProjectMessages(ctx, project.ID, maxContextMessages)
		if err != nil {
			return nil, fmt.Errorf("project messages: %w", err)
		}
		for _, msg := range projectMsgs {
			messages = append(messages, convertDBMessage(msg, emp.ID))
		}
	}

	return trimContext(messages), nil
}

// formatProjectContext builds the initial context message
func formatProjectContext(emp *db.EmployeeWithRole, project *db.Project, task *db.Task) string {
	s := fmt.Sprintf("You are %s, serving as %s in this project.\n\n", emp.Name, emp.RoleTitle)
	s += fmt.Sprintf("## Project: %s\n%s\n\n", project.Name, project.Description)

	if task != nil {
		s += fmt.Sprintf("## Your Current Task: %s\n%s\n\n", task.Title, task.Description)
	}

	s += "## Your Workspace\n"
	s += "You have a personal workspace (your folder in the project's Document/Code tab). Save your work there as you go:\n"
	s += "- Use **save_to_workspace** to save drafts, outlines, research notes, and intermediate code.\n"
	s += "- Use **create_file** for final deliverables to submit to the CEO.\n"
	s += "- Use **execute_command** to run shell commands (build, test, install packages, etc.) in the project workspace.\n\n"
	s += "Please complete your assigned work. Be thorough and professional."

	return s
}

// convertDBMessage converts a DB message to an LLM ChatMessage
func convertDBMessage(msg db.Message, selfEmployeeID string) llm.ChatMessage {
	if msg.SenderType == "agent" && msg.SenderID != nil && *msg.SenderID == selfEmployeeID {
		return llm.ChatMessage{Role: "assistant", Content: msg.Content}
	}
	label := senderLabel(msg)
	return llm.ChatMessage{
		Role:    "user",
		Content: fmt.Sprintf("[%s]: %s", label, msg.Content),
	}
}

func senderLabel(msg db.Message) string {
	switch msg.SenderType {
	case "founder":
		return "Founder"
	case "system":
		return "System"
	default:
		return "Agent"
	}
}

// mergeAndDedup merges two message slices, deduplicates by ID, and sorts chronologically
func mergeAndDedup(a, b []db.Message) []db.Message {
	seen := make(map[string]bool, len(a)+len(b))
	var result []db.Message
	for _, m := range a {
		if !seen[m.ID] {
			seen[m.ID] = true
			result = append(result, m)
		}
	}
	for _, m := range b {
		if !seen[m.ID] {
			seen[m.ID] = true
			result = append(result, m)
		}
	}
	// Sort by CreatedAt
	for i := 1; i < len(result); i++ {
		for j := i; j > 0 && result[j].CreatedAt.Before(result[j-1].CreatedAt); j-- {
			result[j], result[j-1] = result[j-1], result[j]
		}
	}
	return result
}

// trimContext keeps at most maxContextMessages, preserving the first and last N-1 messages
func trimContext(messages []llm.ChatMessage) []llm.ChatMessage {
	if len(messages) <= maxContextMessages {
		return messages
	}
	first := messages[0]
	recent := messages[len(messages)-(maxContextMessages-1):]
	return append([]llm.ChatMessage{first}, recent...)
}
